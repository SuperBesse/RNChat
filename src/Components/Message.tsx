import React from 'react';
import { Text, View, StyleSheet } from 'react-native';
import type { Message as IMEssage } from '@/Types/Message';

type Props = {
  message: IMEssage;
};

const Message = ({ message }: Props) => {
  const formattedDate = (timestamp: number): string => {
    const timestampToDate = new Date(timestamp);
    return timestampToDate.toLocaleTimeString([], {
      hour: '2-digit',
      minute: '2-digit',
    });
  };

  return (
    <View style={styles.container}>
      <View style={styles.messageContainer}>
        <Text style={styles.content}>{message.content}</Text>
      </View>
      <Text style={styles.date}>{formattedDate(message.date)}</Text>
    </View>
  );
};

export default Message;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignItems: 'flex-end',
    paddingHorizontal: 20,
    marginVertical: 10,
  },
  messageContainer: {
    borderTopLeftRadius: 12,
    borderBottomLeftRadius: 12,
    borderBottomRightRadius: 12,
    paddingHorizontal: 10,
    paddingVertical: 12,
    maxWidth: '60%',
    backgroundColor: '#4b9cba',
    marginBottom: 4,
  },
  content: {
    fontSize: 14,
  },
  date: {
    fontSize: 10,
  },
});
