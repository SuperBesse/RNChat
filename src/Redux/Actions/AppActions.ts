export const INIT_APP_START = 'INIT_APP_START'
export const INIT_APP_SUCCESS = 'INIT_APP_SUCCESS'
export const INIT_APP_FAILURE = 'INIT_APP_FAILURE'

export function initApp() {
  return {
    type: INIT_APP_START,
    payload: {},
  }
}
