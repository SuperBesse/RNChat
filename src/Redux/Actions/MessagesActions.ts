export const RECEIVED_NEW_MESSAGE = 'RECEIVED_NEW_MESSAGE';
export const SAVE_NEW_MESSAGE = 'SAVE_NEW_MESSAGE';

import type { Message } from '@/Types/Message';

export function receivedNewMessage(message: Message) {
  return {
    type: RECEIVED_NEW_MESSAGE,
    payload: message,
  };
}
