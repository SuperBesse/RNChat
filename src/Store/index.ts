import AsyncStorage from '@react-native-async-storage/async-storage'
import { persistReducer, persistStore } from 'redux-persist'
import { createStore, applyMiddleware } from 'redux'
import rootReducer from '../Redux/Reducers/index'
import createSagaMiddleware from 'redux-saga'

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whitelist: [],
}

const persistedReducer = persistReducer(persistConfig, rootReducer)
const sagaMiddleware = createSagaMiddleware()

const store = createStore(persistedReducer, applyMiddleware(sagaMiddleware))

const persistor = persistStore(store)

export { store, persistor }
