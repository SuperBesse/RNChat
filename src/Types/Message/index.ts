export type Message = {
  content: string;
  date: number;
  notSent: boolean;
};
