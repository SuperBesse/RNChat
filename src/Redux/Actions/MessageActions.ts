import { Message } from '@/Types/Message';

export const SEND_MESSAGE_START = 'SEND_MESSAGE_START';
export const ADD_MESSAGE_SUCCESS = 'ADD_MESSAGE_SUCCESS';
export const ADD_MESSAGE_FAILURE = 'ADD_MESSAGE_FAILURE';

export function sendMessage(message: String) {
  return {
    type: SEND_MESSAGE_START,
    payload: message,
  };
}

export function addMessageFail(message: Message) {
  return {
    type: ADD_MESSAGE_FAILURE,
    payload: message,
  };
}

export function addMessageSuccess(message: Message) {
  return {
    type: ADD_MESSAGE_SUCCESS,
    payload: message,
  };
}
