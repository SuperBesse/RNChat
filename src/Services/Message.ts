import { Message } from '@/Types/Message';

const randomTime = () => Math.floor(Math.random() * 2000);

const delay = () => new Promise(resolve => setTimeout(resolve, randomTime()));

const randomError = (): Boolean => Math.floor(Math.random() * 100) < 30;

const sendMessage = async (message: Message): Promise<boolean> => {
  console.log('Send Message API: ', message);
  const fakeError: Boolean = randomError();

  if (fakeError) {
    throw new Error('fake error');
  }
  await delay();
  return fakeError;
};

export default sendMessage;
