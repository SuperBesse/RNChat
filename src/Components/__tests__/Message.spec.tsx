import React from 'react';
import { render, screen } from '@testing-library/react-native';
import Message from '../Message';
import { store } from '@/Store';
import { receivedNewMessage } from '@/Redux/Actions/MessagesActions';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

const message = { content: 'test', date: Date.now() };

describe('MessageInput', () => {
  it('should render correctly', () => {
    const tree = render(<Message message={message} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display a text from the message content', async () => {
    render(<Message message={message} />).toJSON();
    const text = screen.getByText('test');
    expect(text).not.toBeNull();
  });
});
