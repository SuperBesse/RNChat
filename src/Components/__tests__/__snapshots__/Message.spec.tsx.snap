// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-end",
      "flexDirection": "column",
      "marginVertical": 10,
      "paddingHorizontal": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#4b9cba",
        "borderBottomLeftRadius": 12,
        "borderBottomRightRadius": 12,
        "borderTopLeftRadius": 12,
        "marginBottom": 4,
        "maxWidth": "60%",
        "paddingHorizontal": 10,
        "paddingVertical": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 14,
        }
      }
    >
      test
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 10,
      }
    }
  >
    00:02
  </Text>
</View>
`;
